var/lib/delphix-platform/os-migration

```diff
@@ -0,0 +1,45 @@
+#!/bin/bash
+#
+# Copyright 2018 Delphix
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+#
+MIGRATE_CONFIG_SCRIPT="/opt/delphix/migration/migrate_config.py"
+MIGRATE_CONFIG_LOG="/var/delphix/migration/log"
+
+function die() {
+	echo "$(basename "$0"): $*" >&2
+	exit 1
+}
+
+function perform_migration() {
+	zpool import -f domain0 || die "Failed to import domain0"
+	# domain0/mds is owned by delphix:staff on Illumos and staff has GID 10.
+	# To minimize the changes in case we need to perform rollback, we add
```

Comment:
Why do we want to minimize changes here? Will we eventually want to change permissions, or will we always keep it set to 10?

I ask because this is likely the best time to make these types of changes, since rollback is available, and we can leverage a checkpoint to revert any changes we make at this time (in the case of a failure). Whereas, if we make changes at some other point, we may not be able to use a checkpoint to handle a failure.

Doing it this way adds another difference between our "normal" linux systems, and our "migrated" linux systems; and I think we should try to avoid these differences.

If the idea is we want to limit what we do on migration, at the cost of creating differences between our "normal" and "migrated" systems, I'm not sure I agree with that approach.

Reply:
We haven't yet decided if we would rely on a checkpoint for the migration upgrade because we do not know how well checkpoint discard / rollback works when we cross the OS boundary.

> If the idea is we want to limit what we do on migration, at the cost of creating differences between our "normal" and "migrated" systems, I'm not sure I agree with that approach.

We could definitely go the other approach that is to make the necessary changes and then have the right tools/scripts to perform a rollback in case things go wrong. 

In this particular case we could chown `domain0/mds` to `postgres:postgres` in the service, and then chown it back to `delphix:staff` in our rollback script.

Reply:
> We haven't yet decided if we would rely on a checkpoint for the migration upgrade because we do not know how well checkpoint discard / rollback works when we cross the OS boundary.

OK. If we can't rely on checkpoints, this will have a large impact on our ability to "rollback" from a failed migration in general; e.g. after MDS migrations have been appliance, and/or ZFS filesystems changes have been made (via the app-gate), we won't be able to rollback.

I wasn't aware of this, so this is good to know.

> We could definitely go the other approach that is to make the necessary changes and then have the right tools/scripts to perform a rollback in case things go wrong.

I just want to be on the same page w.r.t. the approach we're going to take, and collectively agree that's the design we want to go with. If our goal isn't to ensure a "migrated" linux system looks identical to a "normal" linux system, then I won't bring this up anymore; but I think we should be explicit about that decision.

Reply:
> If our goal isn't to ensure a "migrated" linux system looks identical to a "normal" linux system, then I won't bring this up anymore; but I think we should be explicit about that decision.

It will definitely not look identical, as some differences cannot be removed, that said we should strive to make it as close as we can. I still think you brought a good point regardless on whether we make a checkpoint or not. We should definitely optimize the path for migration versus the path for rollback, as long as the changes we make don't render rollback impossible.

As for the checkpoint, it's not "off the table" yet, that said if we do go for it, we would have to discard it early on (i.e. as soon as the stack has booted up).

I'll go ahead with the chown changes.

Reply:
> that said if we do go for it, we would have to discard it early on (i.e. as soon as the stack has booted up).

We do that already. We can't rollback after the stack has passed the "point of no return", which is early in the stack's boot up process. I don't think migration adds any more restrictions (assuming works in general in this case).

> It will definitely not look identical, as some differences cannot be removed, that said we should strive to make it as close as we can.

Right, that's what I meant; as identical as reasonably possible. 

I also think it could be reasonable to make changes to our "normal" linux image to match the requirements of migration (if/when it makes sense); e.g. if we wanted to still go with the "usermod" approach, then we could also run that same command via appliance-build, so "migration" and "normal" linux systems would still be configured similarly.